File: maximum_subarray.py
# -*- coding:utf-8 -*-
'''
Created on 2014-3-18



leetcoder : maximum subarray

Problem : Find the contiguous subarray within an array (containing at least one number) which has the largest sum.

For example, given the array [−2,1,−3,4,−1,2,1,−5,4],
the contiguous subarray [4,−1,2,1] has the largest sum = 6.



@author: xqk
'''

class Solution:
    # @param A, a list of integers
    # @return an integer
    def maxSubArray(self, A):
        if len(A) == 1:
            return A[0];
        maxSum = A[0]
        maxhere = A[0];
        for i in range(len(A) - 1):
            if  maxhere < 0 :
                maxhere = A[i+1];
            else:
                maxhere = maxhere + A[i+1]
            if maxhere > maxSum:
                maxSum = maxhere;
        return maxSum
            
            
                
A = [-2, 1,-3,4,-1,2,1,-5,4]
test = Solution()
print test.maxSubArray(A)

File: Cplusplus/Maximum_Subarray.cpp
/*
Leetcode : Maximum_Subarray.cpp

Find the contiguous subarray within an array (containing at

least one number) which has the largest sum.

For example, given the array [−2,1,−3,4,−1,2,1,−5,4],
the contiguous subarray [4,−1,2,1] has the largest sum = 6.
*/

#include <iostream>
#include <vector>
using namespace std;
class Solution {
public:
    int maxSubArray(int A[], int n) {
        vector<int> arr(n,0);
        int maxV = A[0];
        arr[0] = A[0];
        for(int i = 1 ; i < n ; i++)
        {
            if(arr[i-1] > 0 && arr[i-1] + A[i]>0)
            {
                arr[i] = arr[i-1] + A[i];
            }
            else
            {
                arr[i] = A[i];
            }
            if(arr[i] > maxV)
                maxV = arr[i];

        }
        return maxV;
    }
};

int main()
{
    Solution test = Solution();
    int A[] = {-2,1,-3,4,-1,2,1,-5,4};
    int n = 9;
    cout<<test.maxSubArray(A , n)<<endl;
}
